spring-beans/src/main/java/org/springframework/beans/factory/groovy/GroovyDynamicElementReader.java

```diff
@@ -0,0 +1,145 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.beans.factory.groovy;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.Map;
+
+import groovy.lang.Closure;
+import groovy.lang.GroovyObject;
+import groovy.lang.GroovyObjectSupport;
+import groovy.lang.Writable;
+import groovy.xml.StreamingMarkupBuilder;
+import org.w3c.dom.Element;
+
+import org.springframework.beans.factory.config.BeanDefinitionHolder;
+import org.springframework.beans.factory.support.AbstractBeanDefinition;
+import org.springframework.beans.factory.xml.BeanDefinitionParserDelegate;
+
+/**
+ * Used by GroovyBeanDefinitionReader to read a Spring XML namespace expression
+ * in the Groovy DSL.
+ *
+ * @author Jeff Brown
+ * @author Juergen Hoeller
+ * @since 4.0
+ */
+class GroovyDynamicElementReader extends GroovyObjectSupport {
+
+	private final String rootNamespace;
+
+	private final Map<String, String> xmlNamespaces;
+
+	private final BeanDefinitionParserDelegate delegate;
+
+	private final GroovyBeanDefinitionWrapper beanDefinition;
+
+	protected final boolean decorating;
+
+	private boolean callAfterInvocation = true;
+
+
+	public GroovyDynamicElementReader(String namespace, Map<String, String> namespaceMap,
+			BeanDefinitionParserDelegate delegate, GroovyBeanDefinitionWrapper beanDefinition, boolean decorating) {
+		super();
+		this.rootNamespace = namespace;
+		this.xmlNamespaces = namespaceMap;
+		this.delegate = delegate;
+		this.beanDefinition = beanDefinition;
+		this.decorating = decorating;
+	}
+
+
+	@Override
+	public Object invokeMethod(String name, Object obj) {
+		Object[] args = ((Object[])obj);
+		if (name.equals("doCall")) {
+			@SuppressWarnings("unchecked")
+			Closure<Object> callable = (Closure<Object>) args[0];
+			callable.setResolveStrategy(Closure.DELEGATE_FIRST);
+			callable.setDelegate(this);
+			var result = callable.call();
+
+			if (this.callAfterInvocation) {
+				afterInvocation();
+				this.callAfterInvocation = false;
+			}
+			return result;
+		}
+
+		else {
+			StreamingMarkupBuilder builder = new StreamingMarkupBuilder();
+			var myNamespace = this.rootNamespace;
+			var myNamespaces = this.xmlNamespaces;
+
+			Closure<Object> callable = new Closure<>(this) {
+				@Override
+				public Object call(Object... its) {
+					((GroovyObject) getProperty("mkp")).invokeMethod("declareNamespace", new Object[] {myNamespaces});
```

Comment:
I guess that's an overloaded variant of the `declareNamespace` method that was used in the Groovy version of this class?

Reply:
Ahhhh... I misread the original Groovy code, but I now realize it's in fact the same method and that you simplified the code down to a single invocation of that method instead of the for-each loop in the original Groovy code.